### PR TITLE
[RFC] excalidraw as embedded component

### PR DESCRIPTION
**Pull request by mizchi**
This is proposal and implementation

Example: excalidraw in my (wip) memo app.

or github extension

https://github.com/mizchi/excalidraw-extension

## Goal

I try to use excralidraw as embedded component in my app.

```html
// publish to npm as react component
import { Excalidraw } from "excalidraw";

<MyAppLayout>
  <FooBar />
  <hr />
  <Excalidraw width={800} height={600} onChange={(appState) => {...}}>
</MyAppLayout>
```

but excalidraw requires under window to calc cursor position.

This PR includes delta calculation to root rect.

## What I did

- [x] FixedSideContainer is now on window.
- [x] Cursor position is with left/top of container left/top
- [x] Build to npm library
- [x] Fix fixed layout
- [x] Fix window.onresize on App
- [x] Serializable internalState
- [x] Take initialState props

### CODE REVIEW

File: src/components/App.tsx [hunk 1]
```diff
@@ -759,16 +778,30 @@ export class App extends React.Component<any, AppState> {
...
-      <div className="container">
+      <div
+        className="container"
+        ref={ref => {
+          if (this.containerDeltaInitialized) {
+            return;
+          }
+
+          if (ref) {
+            const rect = ref.getBoundingClientRect();
```
Comment (mizchi) [inline src/components/App.tsx]: This is heavy calculation for layouting in browser so get once.
But it need on resizee again.

File: src/components/App.tsx [hunk 2]
```diff
@@ -565,10 +572,22 @@ export class App extends React.Component<any, AppState> {
...
 
+  private getPositionFromEvent(
+    event:
+      | PointerEvent
+      | MouseEvent
+      | React.MouseEvent<HTMLCanvasElement, MouseEvent>,
+  ) {
+    return [
+      event.screenX - this.containerLeft,
+      event.screenY - this.containerTop,
```
Comment (mizchi) [inline src/components/App.tsx]: I thought it would work, but dose not...

File: src/components/App.tsx [hunk 3]
```diff
@@ -565,10 +572,22 @@ export class App extends React.Component<any, AppState> {
   };
 
   private updateCurrentCursorPosition = (event: MouseEvent) => {
-    cursorX = event.x;
-    cursorY = event.y;
+    cursorX = event.x - this.containerLeft;
+    cursorY = event.y - this.containerTop;
```
Comment (mizchi) [inline src/components/App.tsx]: does not work

**Comment by j-f1** [comment]
It looks like it doesn’t show the mobile layout when it’s rendered too small.

**Comment by mizchi** [comment]
@j-f1 what button? in my pixel it works

**Comment by j-f1** [comment]
I’m saying it should switch to a mobile layout on desktop if it’s too small.

**Comment by mizchi** [comment]
@j-f1 I think you are right about mobile ui but this PR does not include that scope.

File: package.json [hunk 4]
```diff
@@ -34,8 +32,16 @@
     "node-sass": "4.13.1",
     "pepjs": "0.5.2",
     "prettier": "1.19.1",
+    "react": "16.13.0",
+    "react-dom": "16.13.0",
```
Comment (vjeux) [inline package.json]: Why did you move react to devDependencies?

File: public/index.html [hunk 5]
```diff
@@ -151,7 +151,12 @@
     <header>
       <h1 class="visually-hidden">Excalidraw</h1>
     </header>
-    <div id="root"></div>
+
+    <!-- Include padding for container delta check -->
```
Comment (vjeux) [inline public/index.html]: I'm not happy with this change. We should find another solution that doesn't affect the production version.

**Comment by lipis** [comment]
Just out of curiosity, what's the use case for this one?

File: package.json (hunk 4 above)
Comment (mizchi) [reply package.json]: As a component, react and react-dom are required as peerDeps (use instances of host).
If this repo has those, it makes confusing at build step.

**Comment by vjeux** [comment]
I feel like this is something that's worth pursuing but I would like to know more about the use case. Are you trying to embed it in a wiki-like environment where you can edit an image and then save to display it inline? If you tell me more about the use case that would be great!

File: public/index.html (hunk 5 above)
Comment (mizchi) [reply public/index.html]: OK, i will revert this.

**Comment by mizchi** [comment]
My goal is making mdx compiler on browser.

Normal images on markdown is fixed url. I will integrate reshape-able drawing tools on app via mdx syntax.



this `drawing` object will be excaldraw shape and edit with excalidraw editor.

This is my personal goal.

**Comment by vjeux** [comment]
Thanks. That makes a lot of sense! Let’s make it happen then :)

**Comment by vjeux** [comment]
Have you tried embedding the component in an iframe? I feel like this would solve most of the problems without having to change any of the css

**Comment by mizchi** [comment]
Iframe is very limited features to integrate. I need to touch internal state to save on my filesystem. 
I want to save, load and re-edit as excalidraw internal json state. Emitting image is last phase.

If excalidraw works as a component, I can use it lazy load with React.lazy and React.Suspense. Loading payload become better.

(I tried web components's shadowRoot too but it does not work anything. )

**Comment by mizchi** [comment]
@vjeux 

I made excalidraw-chrome extension for real usecase. How about this?

https://github.com/mizchi/excalidraw-extension

**Comment by mizchi** [comment]
I think, in real world use case, excalidraw component needs more features.

- Take initial state as component
- Pluggable UI system on floating layer
- Emit callback to internal shapes
- Selectable storage adapters

**Comment by dwelle** [comment]
Note that this PR regressed some UI elements on desktop. The tools and zoom buttons are bunched up to top-left corner:

image

**Comment by mizchi** [comment]
@dwelle Oops, I will fix them with `position:relative/absolute`

**Comment by dwelle** [comment]
The Island/App margins still differ from production. @j-f1 might wanna review this.

**Comment by mizchi** [comment]
I guess

**Comment by mizchi** [comment]
I added padding to menus

File: src/styles.scss [hunk 6]
```diff
@@ -1,371 +1,365 @@
 @import "./theme.css";
 @import "./_variables";
 
-body {
-  margin: 0;
+.excalidraw {
```
Comment (mizchi) [inline src/styles.scss]: As a component I need to enclose other global selectors.

File: src/app.scss [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+@import "./_variables";
+
+body {
+  background-color: #eee;
+  margin: 0;
+  -webkit-text-size-adjust: 100%;
+  user-select: none;
+  width: 100vw;
+  height: 100vh;
+}
```
Comment (mizchi) [inline src/app.scss]: Override host body style when I use it on my app.

File: src/styles.scss (hunk 6 above)
Comment (mizchi) [reply src/styles.scss]: ... so most changes in this file are indent

Comment (mizchi) [reply src/styles.scss]: or should I use shadowRoot? (it drop IE)

File: src/index.tsx [hunk 8]
```diff
@@ -1,29 +1,68 @@
...
+    return () => window.removeEventListener("resize", onResize);
+  }, []);
+  return (
+    <Excalidraw
+      width={size.width}
+      height={size.height}
+      initialState={props.initialState || undefined}
+      onChange={internalState => {
+        saveToLocalStorage(internalState.elements, internalState.appState);
+      }}
```
Comment (mizchi) [inline src/index.tsx]: This refactor can change storage backend as you like.

**Comment by mizchi** [comment]
My RFC commits are all now.

**Comment by mizchi** [comment]
It's difficult to maintain so I quit this branch. Sorry, I tried big changes at first.

But by my dog fooding to integrate my app, I can understand what I should do by step.

---

If this feature are needed, I need these steps.

## State serialization

- Define `InternalState`  and make it `Serializable/Deserialize` (Map => Object)
- Handle state initialization/restoration from outside of App component and emit it by onChange.

## Position calculation

- Consider container Left/Top for position calculation.

## Isolation

- CSS: fixed to relative/absolute
- CSS: Scoped css (shadowRoot or top)

@vjeux I will try state serialization for restart. How do you think?

**Comment by dwelle** [comment]
> It's difficult to maintain so I quit this branch. Sorry, I tried big changes at first.

Unclear what's difficult to maintain. We've had some bigger refactors lately. I was about to offer help in conflict resolution.

We haven't had much time to look at this PR yet because it's huge as you say, and other things took priority.

**Comment by lipis** [comment]
I think we can close this for now, and see in the future if we come back to it. Thanks for your time and ideas.
